File: src/jsTest/kotlin/com/scoreo/infrastructure/google/GoogleAuthServiceTest.kt
package com.scoreo.infrastructure.google

import com.scoreo.domain.port.SyncException
import kotlin.js.js
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest

/**
 * Tests for GoogleAuthService
 *
 * Tests OAuth login, token management, token refresh, and logout flows.
 * Covers happy path and error scenarios (GIS not loaded, network errors, auth failures).
 *
 * GoogleAuthService uses Google Identity Services which loads async in browser.
 * Tests verify contract behavior for all scenarios.
 *
 * Coverage:
 * - Token storage (accessToken, expiresAt, idToken)
 * - Login flow (async callback handling)
 * - Refresh token flow (silent refresh with prompt="")
 * - Logout (token revocation)
 * - Error handling (GIS unavailable, auth errors)
 * - Multiple operations (lifecycle tests)
 */
class GoogleAuthServiceTest {

    // ── Initialization ──

    @Test
    fun `service initializes with null accessToken`() {
        val service = GoogleAuthService()
        
        assertNull(service.accessToken)
    }

    @Test
    fun `service initializes with null expiresAt`() {
        val service = GoogleAuthService()
        
        assertNull(service.expiresAt)
    }

    @Test
    fun `service initializes with null idToken`() {
        val service = GoogleAuthService()
        
        assertNull(service.idToken)
    }

    // ── Token Storage ──

    @Test
    fun `accessToken can be set and retrieved`() {
        val service = GoogleAuthService()
        service.accessToken = "test-token-123"
        
        assertEquals("test-token-123", service.accessToken)
    }

    @Test
    fun `expiresAt can be set and retrieved`() {
        val service = GoogleAuthService()
        val expiryTime = (js("Date.now()") as Double).toLong() + 3600_000L
        service.expiresAt = expiryTime
        
        assertEquals(expiryTime, service.expiresAt)
    }

    @Test
    fun `idToken can be set and retrieved`() {
        val service = GoogleAuthService()
        service.idToken = "id-token-xyz"
        
        assertEquals("id-token-xyz", service.idToken)
    }

    @Test
    fun `all tokens can be set independently`() {
        val service = GoogleAuthService()
        service.accessToken = "access-123"
        service.idToken = "id-456"
        service.expiresAt = 9999000000000L
        
        assertEquals("access-123", service.accessToken)
        assertEquals("id-456", service.idToken)
        assertEquals(9999000000000L, service.expiresAt)
    }

    // ── Login Flow ──

    @Test
    fun `login accepts clientId parameter`() = runTest {
        val service = GoogleAuthService()
        var callbackInvoked = false
        
        service.login("test-client-id", "profile email", onResult = { result ->
            callbackInvoked = true
        })
        
        kotlinx.coroutines.delay(100)
        
        // Should not throw
        assertTrue(true)
    }

    @Test
    fun `login accepts scope parameter`() = runTest {
        val service = GoogleAuthService()
        
        service.login("client-id", "profile email openid", onResult = { result ->
            // Scope should be passed to GIS
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `login callback receives Result`() = runTest {
        val service = GoogleAuthService()
        var resultReceived: Result<String>? = null
        
        service.login("client-id", "profile", onResult = { result ->
            resultReceived = result
        })
        
        kotlinx.coroutines.delay(100)
        
        // Callback should be invoked eventually (GIS might not be available)
        assertTrue(true)
    }

    @Test
    fun `login retries if GIS not immediately available`() = runTest {
        val service = GoogleAuthService()
        var resultReceived: Result<String>? = null
        
        service.login("client-id", "profile", onResult = { result ->
            resultReceived = result
        })
        
        // Wait for retry loop
        kotlinx.coroutines.delay(2500)
        
        // Should eventually get a result (success or GIS not available error)
        assertNotNull(resultReceived)
    }

    // ── Refresh Flow ──

    @Test
    fun `refreshToken accepts clientId parameter`() = runTest {
        val service = GoogleAuthService()
        
        service.refreshToken("client-id", "profile", onResult = { result ->
            // Should attempt refresh
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `refreshToken accepts scope parameter`() = runTest {
        val service = GoogleAuthService()
        
        service.refreshToken("client-id", "profile email", onResult = { result ->
            // Scope should be passed to GIS
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `refreshToken callback receives Result`() = runTest {
        val service = GoogleAuthService()
        var resultReceived: Result<String>? = null
        
        service.refreshToken("client-id", "profile", onResult = { result ->
            resultReceived = result
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `refreshToken uses silent refresh with prompt empty`() = runTest {
        val service = GoogleAuthService()
        
        service.refreshToken("client-id", "profile", onResult = { result ->
            // Should use prompt="" for silent refresh
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `refreshToken retries if GIS not available`() = runTest {
        val service = GoogleAuthService()
        var resultReceived: Result<String>? = null
        
        service.refreshToken("client-id", "profile", onResult = { result ->
            resultReceived = result
        })
        
        // Wait for retry loop
        kotlinx.coroutines.delay(2500)
        
        // Should eventually resolve
        assertNotNull(resultReceived)
    }

    // ── Logout ──

    @Test
    fun `logout clears accessToken`() {
        val service = GoogleAuthService()
        service.accessToken = "token"
        
        service.logout()
        
        assertNull(service.accessToken)
    }

    @Test
    fun `logout clears expiresAt`() {
        val service = GoogleAuthService()
        service.expiresAt = 999999999L
        
        service.logout()
        
        assertNull(service.expiresAt)
    }

    @Test
    fun `logout when no token is safe`() {
        val service = GoogleAuthService()
        assertNull(service.accessToken)
        
        service.logout()
        
        assertNull(service.accessToken)
    }

    @Test
    fun `logout after already logged out is safe`() {
        val service = GoogleAuthService()
        service.accessToken = "token"
        
        service.logout()
        assertNull(service.accessToken)
        
        service.logout()
        assertNull(service.accessToken)
    }

    // ── Token Expiry Calculation ──

    @Test
    fun `expiresAt represents future time`() {
        val service = GoogleAuthService()
        val now = (js("Date.now()") as Double).toLong()
        service.expiresAt = now + 3600_000L  // 1 hour from now
        
        assertTrue((service.expiresAt ?: 0) > now)
    }

    @Test
    fun `expiresAt calculation matches expires_in conversion`() {
        val service = GoogleAuthService()
        val before = (js("Date.now()") as Double).toLong()
        val expiresIn = 3600  // 1 hour in seconds
        service.expiresAt = before + (expiresIn * 1000L)
        val after = (js("Date.now()") as Double).toLong()
        
        val expiresAt = service.expiresAt ?: 0
        
        // Should be approximately before + 1 hour
        assertTrue(expiresAt >= before + (expiresIn * 1000L) - 100)
        assertTrue(expiresAt <= after + (expiresIn * 1000L) + 100)
    }

    // ── Multiple Operations ──

    @Test
    fun `service instance independence`() {
        val service1 = GoogleAuthService()
        val service2 = GoogleAuthService()
        
        service1.accessToken = "token-1"
        service2.accessToken = "token-2"
        
        assertEquals("token-1", service1.accessToken)
        assertEquals("token-2", service2.accessToken)
    }

    @Test
    fun `service state cleared after logout`() {
        val service = GoogleAuthService()
        service.accessToken = "token"
        service.expiresAt = 999999999L
        service.idToken = "id-token"
        
        service.logout()
        
        assertNull(service.accessToken)
        assertNull(service.expiresAt)
    }

    @Test
    fun `multiple logouts are safe`() {
        val service = GoogleAuthService()
        service.accessToken = "token"
        
        service.logout()
        service.logout()
        service.logout()
        
        assertNull(service.accessToken)
    }

    // ── Scope Handling ──

    @Test
    fun `login accepts standard OAuth scopes`() = runTest {
        val service = GoogleAuthService()
        
        service.login("client-id", "profile", onResult = { result ->
            // Should accept profile scope
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `login accepts email scope`() = runTest {
        val service = GoogleAuthService()
        
        service.login("client-id", "email", onResult = { result ->
            // Should accept email scope
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `login accepts multiple scopes`() = runTest {
        val service = GoogleAuthService()
        
        service.login("client-id", "profile email openid", onResult = { result ->
            // Should accept multiple scopes
        })
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }

    @Test
    fun `login accepts Drive API scope`() = runTest {
        val service = GoogleAuthService()
        
        service.login(
            "client-id",
            "https://www.googleapis.com/auth/drive.appdata",
            onResult = { result ->
                // Should accept Drive scope URL
            }
        )
        
        kotlinx.coroutines.delay(100)
        
        assertTrue(true)
    }
}

File: apps/scoreo/src/infrastructure/google/googleAuthService.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import type { SyncException } from '../../domain/port/cloudSyncRepository'
import type { Result } from '../../domain/result'
import { GoogleAuthService } from './googleAuthService'

/**
 * The Kotlin jsTest suite for GoogleAuthService mostly asserts `true` after calling
 * login/refreshToken, because GIS is never actually loaded in that test environment —
 * it only exercises the "not loaded" retry path. Here we mock `window.google.accounts.oauth2`
 * so the tests can assert real behavior (token storage, error propagation, silent refresh)
 * instead of tautologies.
 */
function installMockGis(overrides: {
  onRequestAccessToken?: (config: {
    client_id: string
    scope: string
    hint?: string
    callback: (r: { access_token: string; expires_in: number; token_type: string }) => void
    error_callback: (e: { type: string; message?: string }) => void
  }, overrideConfig?: { prompt?: string }) => void
  onRevoke?: (token: string) => void
} = {}) {
  window.google = {
    accounts: {
      oauth2: {
        initTokenClient: (config) => ({
          requestAccessToken: (overrideConfig) => overrides.onRequestAccessToken?.(config, overrideConfig),
        }),
        revoke: (token, callback) => {
          overrides.onRevoke?.(token)
          callback()
        },
      },
    },
  }
}

/**
 * `login` resolves the account email through the OIDC userinfo endpoint before invoking its
 * callback — GIS's token model never returns an id_token, so this is the only way to learn
 * which account was authorized (and hence what to pass back as `hint` on a silent refresh).
 */
function mockUserinfo(response: { ok: true; email: string } | { ok: false }) {
  const fetchMock = vi.fn(async () =>
    response.ok
      ? ({ ok: true, json: async () => ({ email: response.email }) } as Response)
      : ({ ok: false, status: 401, json: async () => ({}) } as Response),
  )
  vi.stubGlobal('fetch', fetchMock)
  return fetchMock
}

/** Resolves once GIS has called back, so tests can await login's asynchronous email lookup. */
function loginAsync(service: GoogleAuthService, clientId = 'client-id', scope = 'openid email') {
  return new Promise<Result<string, SyncException>>((resolve) => service.login(clientId, scope, resolve))
}

describe('GoogleAuthService', () => {
  beforeEach(() => {
    vi.useFakeTimers()
    mockUserinfo({ ok: true, email: 'player@example.com' })
  })

  afterEach(() => {
    delete window.google
    vi.useRealTimers()
    vi.unstubAllGlobals()
  })

  it('initializes with null tokens', () => {
    const service = new GoogleAuthService()
    expect(service.accessToken).toBeNull()
    expect(service.expiresAt).toBeNull()
  })

  it('login stores accessToken and expiresAt on success', async () => {
    installMockGis({
      onRequestAccessToken: (config) => {
        config.callback({ access_token: 'token-abc', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    const result = await loginAsync(service)

    expect(service.accessToken).toBe('token-abc')
    expect(service.expiresAt).toBeGreaterThan(Date.now())
    expect(result).toEqual({ ok: true, value: 'token-abc' })
  })

  it('login resolves the account email from the userinfo endpoint', async () => {
    const fetchMock = mockUserinfo({ ok: true, email: 'player@example.com' })
    installMockGis({
      onRequestAccessToken: (config) => {
        config.callback({ access_token: 'token-abc', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    await loginAsync(service)

    expect(service.accountEmail).toBe('player@example.com')
    expect(fetchMock).toHaveBeenCalledWith(
      'https://openidconnect.googleapis.com/v1/userinfo',
      expect.objectContaining({ headers: { Authorization: 'Bearer token-abc' } }),
    )
  })

  it('login still succeeds when the userinfo lookup fails, leaving accountEmail null', async () => {
    mockUserinfo({ ok: false })
    installMockGis({
      onRequestAccessToken: (config) => {
        config.callback({ access_token: 'token-abc', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    const result = await loginAsync(service)

    expect(result).toEqual({ ok: true, value: 'token-abc' })
    expect(service.accessToken).toBe('token-abc')
    expect(service.accountEmail).toBeNull()
  })

  it('login still succeeds when the userinfo call rejects outright', async () => {
    vi.stubGlobal('fetch', vi.fn(async () => { throw new Error('offline') }))
    installMockGis({
      onRequestAccessToken: (config) => {
        config.callback({ access_token: 'token-abc', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    const result = await loginAsync(service)

    expect(result).toEqual({ ok: true, value: 'token-abc' })
    expect(service.accountEmail).toBeNull()
  })

  it('login calls initTokenClient with a client_id key, matching the real GIS API contract', () => {
    let capturedConfig: Record<string, unknown> | undefined
    window.google = {
      accounts: {
        oauth2: {
          initTokenClient: (config) => {
            capturedConfig = config as unknown as Record<string, unknown>
            return { requestAccessToken: () => {} }
          },
          revoke: (_token, callback) => callback(),
        },
      },
    }
    const service = new GoogleAuthService()

    service.login('client-id-value', 'openid email', vi.fn())

    expect(capturedConfig?.client_id).toBe('client-id-value')
    expect(capturedConfig?.clientId).toBeUndefined()
  })

  it('login surfaces NotAuthenticated on GIS error_callback', () => {
    installMockGis({
      onRequestAccessToken: (config) => {
        config.error_callback({ type: 'popup_closed', message: 'User closed the popup' })
      },
    })
    const service = new GoogleAuthService()
    const onResult = vi.fn()

    service.login('client-id', 'openid email', onResult)

    expect(onResult).toHaveBeenCalledWith({
      ok: false,
      error: { kind: 'NotAuthenticated', message: 'User closed the popup' },
    })
  })

  it('login retries until GIS becomes available', async () => {
    const service = new GoogleAuthService()
    const onResult = vi.fn()
    const result = new Promise((resolve) => service.login('client-id', 'openid email', (r) => { onResult(r); resolve(r) }))

    expect(onResult).not.toHaveBeenCalled()

    installMockGis({
      onRequestAccessToken: (config) => {
        config.callback({ access_token: 'late-token', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    vi.advanceTimersByTime(200)

    await expect(result).resolves.toEqual({ ok: true, value: 'late-token' })
  })

  it('login fails with NotAuthenticated after exhausting retries when GIS never loads', () => {
    const service = new GoogleAuthService()
    const onResult = vi.fn()

    service.login('client-id', 'openid email', onResult)
    vi.advanceTimersByTime(10 * 200)

    expect(onResult).toHaveBeenCalledWith({
      ok: false,
      error: { kind: 'NotAuthenticated', message: 'Google Identity Services not loaded' },
    })
  })

  it('refreshToken uses a silent prompt', () => {
    let capturedOverride: { prompt?: string } | undefined
    installMockGis({
      onRequestAccessToken: (config, overrideConfig) => {
        capturedOverride = overrideConfig
        config.callback({ access_token: 'refreshed-token', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    service.refreshToken('client-id', 'openid email', vi.fn())

    expect(capturedOverride).toEqual({ prompt: '' })
    expect(service.accessToken).toBe('refreshed-token')
  })

  it('refreshToken passes the hint to initTokenClient so GIS picks the account without asking', () => {
    let capturedConfig: { hint?: string } | undefined
    let capturedOverride: { prompt?: string } | undefined
    installMockGis({
      onRequestAccessToken: (config, overrideConfig) => {
        capturedConfig = config
        capturedOverride = overrideConfig
        config.callback({ access_token: 'refreshed-token', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    service.refreshToken('client-id', 'openid email', vi.fn(), 'player@example.com')

    expect(capturedConfig?.hint).toBe('player@example.com')
    expect(capturedOverride).toEqual({ prompt: '' })
  })

  it('refreshToken omits hint entirely when none is known', () => {
    let capturedConfig: Record<string, unknown> | undefined
    installMockGis({
      onRequestAccessToken: (config) => {
        capturedConfig = config as unknown as Record<string, unknown>
        config.callback({ access_token: 'refreshed-token', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    service.refreshToken('client-id', 'openid email', vi.fn())

    expect(capturedConfig).not.toHaveProperty('hint')
  })

  it('refreshToken does not re-query userinfo (the email is already persisted)', () => {
    const fetchMock = mockUserinfo({ ok: true, email: 'player@example.com' })
    installMockGis({
      onRequestAccessToken: (config) => {
        config.callback({ access_token: 'refreshed-token', expires_in: 3600, token_type: 'Bearer' })
      },
    })
    const service = new GoogleAuthService()

    service.refreshToken('client-id', 'openid email', vi.fn(), 'player@example.com')

    expect(fetchMock).not.toHaveBeenCalled()
  })

  it('refreshToken propagates errors', () => {
    installMockGis({
      onRequestAccessToken: (config) => {
        config.error_callback({ type: 'server_error' })
      },
    })
    const service = new GoogleAuthService()
    const onResult = vi.fn()

    service.refreshToken('client-id', 'openid email', onResult)

    expect(onResult).toHaveBeenCalledWith({
      ok: false,
      error: { kind: 'NotAuthenticated', message: 'Authentication failed (server_error)' },
    })
  })

  it('logout revokes the token and clears state', () => {
    let revokedToken: string | undefined
    installMockGis({ onRevoke: (token) => (revokedToken = token) })
    const service = new GoogleAuthService()
    service.accessToken = 'token-to-revoke'
    service.expiresAt = Date.now() + 1000

    service.logout()

    expect(revokedToken).toBe('token-to-revoke')
    expect(service.accessToken).toBeNull()
    expect(service.expiresAt).toBeNull()
    expect(service.accountEmail).toBeNull()
  })

  it('logout is safe when there is no token', () => {
    installMockGis()
    const service = new GoogleAuthService()

    expect(() => service.logout()).not.toThrow()
    expect(service.accessToken).toBeNull()
  })

  it('service instances are independent', () => {
    const service1 = new GoogleAuthService()
    const service2 = new GoogleAuthService()

    service1.accessToken = 'token-1'
    service2.accessToken = 'token-2'

    expect(service1.accessToken).toBe('token-1')
    expect(service2.accessToken).toBe('token-2')
  })
})
